src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -1149,23 +1150,84 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
             return null;
         }
 
-        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node)
+        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node, InterpolatedStringHandlerData? data)
         {
-            foreach (var expr in node.Parts)
+            // If there can be any branching, then we need to treat the expressions
+            // as optionally evaluated. Otherwise, we treat them as always evaluated
+            switch (data)
             {
-                VisitRvalue(expr);
+                case null:
+                    visitParts();
+                    break;
+                case { HasTrailingHandlerValidityParameter: false, UsesBoolReturns: false, Construction: var construction }:
+                    VisitRvalue(construction);
+                    visitParts();
+                    break;
+                case { UsesBoolReturns: var usesBoolReturns, HasTrailingHandlerValidityParameter: var hasTrailingValidityParameter, Construction: var construction }:
+                    VisitRvalue(construction);
+
+                    if (node.Parts.IsEmpty)
+                    {
+                        break;
+                    }
+
+                    TLocalState beforePartsState;
+                    ReadOnlySpan<BoundExpression> remainingParts;
+
+                    if (hasTrailingValidityParameter)
+                    {
+                        beforePartsState = State.Clone();
+                        remainingParts = node.Parts.AsSpan();
+                    }
+                    else
+                    {
+                        Visit(node.Parts[0]);
+                        beforePartsState = State.Clone();
+                        remainingParts = node.Parts.AsSpan()[1..];
+                    }
+
+                    foreach (var expr in remainingParts)
+                    {
+                        VisitRvalue(expr);
+                        if (usesBoolReturns)
+                        {
+                            Join(ref beforePartsState, ref State);
+                        }
+                    }
+
+                    if (usesBoolReturns)
+                    {
+                        // Already been joined after the last part, just assign
+                        State = beforePartsState;
+                    }
+                    else
+                    {
+                        Join(ref State, ref beforePartsState);
```

Comment:
I didn't understand this case. Why do we need to `Join` when none of the evaluations are conditional?

Reply:
It is not. `hasTrailingValidityParameter` will be true in that case.

Reply:
Thanks for the clarification

Reply:
Nit: this also confused me. Consider adding a comment.